### PR TITLE
[fix]curvefs: enableSumInDir in multi mount situation

### PR DESCRIPTION
**Pull request by bit-dance**
### What problem does this PR solve?

Issue Number: #2094

Problem Summary:

### What is changed and how it works?
When the fs mount is mounted multiple times with different enableSumInDir switches, the xattr information recorded in the metadata will be inaccurate.
What's Changed:
check the mds flag in lease_executor and update for fuse client
How it Works:
just like above
Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by bit-dance** [comment]
cicheck

File: curvefs/src/mds/fs_manager.cpp [hunk 1]
```diff
@@ -843,6 +843,18 @@ void FsManager::RefreshSession(const RefreshSessionRequest* request,
...
+    FsInfoWrapper wrapper;
+    FSStatusCode ret = fsStorage_->Get(request->fsname(), &wrapper);
+    if (ret != FSStatusCode::OK) {
+        LOG(WARNING) << "GetFsInfo fail, get fs fail, fsName = "
+                    << request->fsname()
+                    << ", errCode = " << FSStatusCode_Name(ret);
+        return;
+    }
+
+    *fsinfo = wrapper.ProtoFsInfo();
```
Comment (SeanHai) [inline curvefs/src/mds/fs_manager.cpp:856]: *fsinfo not initialized, here temporary variable fsinfo here is not necessary
`response->set_enablesumindir(wrapper.ProtoFsInfo().enablesumindir());` or Add method GetEnableSumInDir in class FsInfoWrapper and can use directly like `response->set_enablesumindir(wrapper.GetEnableSumInDir())`

File: curvefs/src/client/rpcclient/mds_client.cpp [hunk 2]
```diff
@@ -532,6 +533,7 @@ MdsClientImpl::RefreshSession(const std::vector<PartitionTxId> &txIds,
             LOG(INFO) << "RefreshSession need update partition txid list: "
                       << response.DebugString();
         }
+        *enableSumInDir = response.enablesumindir();
```
Comment (SeanHai) [inline curvefs/src/client/rpcclient/mds_client.cpp:536]: There may be thread safety issues here.

**Comment by bit-dance** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Review by Ziy1-Tan — COMMENTED** [review]
`if(*enableSumInDir && !response.enablesumindir())` more clearly

**Comment by bit-dance** [comment]
> `if(*enableSumInDir && !response.enablesumindir())` more clearly

Thx!And i will update in new version later.

File: curvefs/src/client/rpcclient/mds_client.cpp [hunk 3]
```diff
@@ -532,6 +533,9 @@ MdsClientImpl::RefreshSession(const std::vector<PartitionTxId> &txIds,
             LOG(INFO) << "RefreshSession need update partition txid list: "
                       << response.DebugString();
         }
+        if (*enableSumInDir == true && response.enablesumindir() == false) {
+            *enableSumInDir = response.enablesumindir();
```
Comment (SeanHai) [inline curvefs/src/client/rpcclient/mds_client.cpp:537]: use enableSumInDir->store(response.enablesumindir());

And change use enableSumInDir_ value to enableSumInDir_.load()

std::atomic::store and std::atomic::load is thread safety.

Comment (bit-dance) [reply curvefs/src/client/rpcclient/mds_client.cpp:537]: Get it! I will change afternoon.

**Comment by bit-dance** [comment]
cicheck

File: curvefs/src/client/lease/lease_excutor.h [hunk 4]
```diff
@@ -44,8 +45,10 @@ namespace client {
 class LeaseExecutor : public LeaseExecutorBase {
  public:
     LeaseExecutor(const LeaseOpt &opt, std::shared_ptr<MetaCache> metaCache,
-                  std::shared_ptr<MdsClient> mdsCli)
-        : opt_(opt), metaCache_(metaCache), mdsCli_(mdsCli) {}
+                  std::shared_ptr<MdsClient> mdsCli,
+                  std::atomic<bool>* enableSumInDir = nullptr)
```
Comment (bit-dance) [inline curvefs/src/client/lease/lease_excutor.h:49]: Here will be nullptr exception,should remove default assignment.

File: curvefs/test/client/rpcclient/mds_client_test.cpp [hunk 5]
```diff
@@ -837,7 +838,7 @@ TEST_F(MdsClientImplTest, RefreshSession) {
 
     // out
     std::vector<PartitionTxId> out;
-
+    std::atomic<bool>* enableSumInDir = new std::atomic<bool> (true);
```
Comment (bit-dance) [inline curvefs/test/client/rpcclient/mds_client_test.cpp:841]: std::atomic<bool> enableSumInDir;
enableSumInDir.store(false);

**Comment by bit-dance** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Comment by SeanHai** [comment]
cicheck

**Review by SeanHai — APPROVED** [review]

**Comment by SeanHai** [comment]
LGTM

File: curvefs/src/client/fuse_client.cpp [hunk 6]
```diff
@@ -1117,7 +1118,7 @@ CURVEFS_ERROR FuseClient::FuseOpSetAttr(fuse_req_t req, fuse_ino_t ino,
         inodeWrapper->GetInodeAttrLocked(&inodeAttr);
         InodeAttr2ParamAttr(inodeAttr, attrOut);
 
-        if (enableSumInDir_ && changeSize != 0) {
+        if (enableSumInDir_.load()&& changeSize != 0) {
```
Comment (Ziy1-Tan) [inline curvefs/src/client/fuse_client.cpp:1121]:
```suggestion
        if (enableSumInDir_.load() && changeSize != 0) {
```

File: curvefs/src/mds/fs_info_wrapper.cpp [hunk 7]
```diff
@@ -113,6 +113,10 @@ void FsInfoWrapper::AddMountPoint(const Mountpoint& mp) {
     *p = mp;
 
     fsInfo_.set_mountnum(fsInfo_.mountnum() + 1);
+
+    if (fsInfo_.enablesumindir()&&fsInfo_.mountnum() > 1) {
```
Comment (Ziy1-Tan) [inline curvefs/src/mds/fs_info_wrapper.cpp:117]:
```suggestion
    if (fsInfo_.enablesumindir() && fsInfo_.mountnum() > 1) {
```

**Review by Ziy1-Tan — CHANGES_REQUESTED** [review]

File: curvefs/src/mds/fs_info_wrapper.cpp (hunk 7 above)
Comment (bit-dance) [reply curvefs/src/mds/fs_info_wrapper.cpp:117]: Thx!And i will update later

**Comment by Ziy1-Tan** [comment]
cicheck

**Comment by bit-dance** [comment]
cicheck

**Review by Ziy1-Tan — APPROVED** [review]

**Pull request merged**